### PR TITLE
Virtual tutor dev merge

### PR DESCRIPTION
**Pull request by pernilje**
## Description

This pull request contains these completed PBI's:

- [Web Create] As a content creator, I want to be able to put information on the new section of the new course I'm creating.
- [Web Create] As a content creator, I want to be able to put general information on the new courses I'm creating.
    -  This PBI is almost completely done, but you still can't upload a cover image.
- [Web Create] As a content creator, I want to be able to add a section on the new course I'm creating.
- [Web Edit] As a content creator, I want to be able to edit the sections of the courses I haven't published yet.
- [Web Edit] As a content creator, I want to be able to edit the general info of the courses I haven't published yet

We have modified already existing files to show the proper pages and remove multiple errors from each page. 

## Changes

- Fixed a lot of errors and hard crashes that inhibited multiple pages from being accessed.

- In course creation.
  - Added button for cover image upload. (not functional)
  - Design made from style guide.
  - Translated to Portuguese.
  - Added categories, level, and estimated time.
  
- In the course edit page.
  - You can now update the information in the course with the course update button.
  - You can now create sections for the course.
  - You can now delete the course.
  
- In view courses page.
  - You can now view and access all the courses in the database.
  - Merged together with the new design.

- In the section editing page.
  - You can now change the title and description of a section.
  - You can now delete a section.
  - The update section button now saves and updates the information on the section.
  - Pop up buttons to access the create lecture or exercise form.

## Checklist

- [x] Code has been tested locally and passes all relevant tests.
- [x] Documentation has been updated to reflect the changes, if applicable.
- [x] Code follows the established coding style and guidelines of the project.
- [x] All new and existing tests related to the changes have passed.
- [x] Any necessary dependencies or new packages have been properly documented.
- [x] Pull request title and description are clear and descriptive.
- [x] Reviewers have been assigned to the pull request.
- [x] Any potential security implications have been considered and addressed.
- [x] Performance impact of the changes has been evaluated, if relevant.

## Notes for Reviewers
The backend used for this pull requist is "VirtualTurtor-Merge".
This code needs access to mongoDB to function.

Linting gives a lot of errors, we fixed a lot of the existing errors linting gave, but the rest are non trivial errors that don't inhibit the code and are therefore not a priority right now.

### CODE REVIEW

**Comment by MaizeHH** [comment]
Would you please provide which backend branch this runs on and other potential prerequisites? 😄

File: app/src/pages/SectionEdit.tsx
Comment (JackSMunn) [inline app/src/pages/SectionEdit.tsx:83]: This way of filtering out errors, does it work? Or are there any status where you would get something that is not correct?

Comment (JackSMunn) [reply app/src/pages/SectionEdit.tsx:83]: I don't know...

File: app/src/pages/SectionEdit.tsx
Comment (JackSMunn) [inline app/src/pages/SectionEdit.tsx:125]: dead code?

File: app/src/services/course.services.ts
Comment (JackSMunn) [inline app/src/services/course.services.ts:12]: If you made changes to the model, please write it into the MODELS.md

File: app/src/services/course.services.ts
Comment (JackSMunn) [inline app/src/services/course.services.ts:21]: console log needed here?

File: app/src/services/course.services.ts
Comment (JackSMunn) [inline app/src/services/course.services.ts:29]: If you made changes to the model, please write it into the MODELS.md

File: app/src/services/course.services.ts
Comment (JackSMunn) [inline app/src/services/course.services.ts:39]: Do not have actions in the endpoint. Also endpoints should be pluralized (courses instead of course)

This would be api/courses/coverImage/:id

I also believe, if I'm not mistaken, that we should not use camelcase in endpoints. @Jatewo @Havkost, do you guys remember?

Comment (JackSMunn) [reply app/src/services/course.services.ts:39]: Also, should we use PATCH instead of POST?

File: app/src/services/course.services.ts
Comment (JackSMunn) [inline app/src/services/course.services.ts:76]: PATCH instead of POST i believe.

File: app/src/services/course.services.ts
Comment (JackSMunn) [inline app/src/services/course.services.ts:78]: Is this todo done?

File: app/src/services/course.services.ts
Comment (JackSMunn) [inline app/src/services/course.services.ts:94]: console log needed?

File: app/src/services/course.services.ts
Comment (JackSMunn) [inline app/src/services/course.services.ts:96]: Do not have actions in the endpoint.
This would be api/courses/coverImage/:id

A thing to note, that I am not sure on, should cover-image be pluralized?

File: app/src/services/exercise.services.ts
Comment (JackSMunn) [inline app/src/services/exercise.services.ts:12]: Do not have actions in the endpoint. 

Should be api/exercises/:id

File: app/src/services/exercise.services.ts
Comment (JackSMunn) [inline app/src/services/exercise.services.ts:21]: Do not have actions in the endpoint

should be api/exercises/:id

Comment (JackSMunn) [reply app/src/services/exercise.services.ts:21]: Altso, should it be PATCH instead of POST?

File: app/src/services/lecture.services.ts
Comment (JackSMunn) [inline app/src/services/lecture.services.ts:17]: dead code?

File: app/src/services/lecture.services.ts
Comment (JackSMunn) [inline app/src/services/lecture.services.ts:19]: Do not have actions in endpoint

should be api/lectures/:id

Also should it be PUT instead of POST? Any thoughts

File: app/src/services/section.services.ts
Comment (JackSMunn) [inline app/src/services/section.services.ts:40]: Do not have actions in endpoint

should be api/sections/:id

Should it be PUT instead of POST?

File: app/src/services/section.services.ts
Comment (JackSMunn) [inline app/src/services/section.services.ts:49]: Do not have actions in endpoints

should be api/sections/:id

**Review by JackSMunn — CHANGES_REQUESTED** [review]
GJ! 👍 Karina you owe me cake

I can approve when it is tested locally (on my end), but that will be tomorrow.

The whole frontend lacks testing.

I have a few things for you to address. Mainly being dead code, endpoint guidelines and changes to models.

Please address the comments, if you do not think changes are needed, please do comment why and we can resolve changes faster.

File: app/src/components/CreateLecturePopUp.tsx
Comment (magnemo) [reply app/src/components/CreateLecturePopUp.tsx:46]: Not dead, but when tokens are done dummy token should be removed and useToken should be uncomment. We have add a todo comment here to explain this

File: app/src/components/CreateLecturePopUp.tsx
Comment (magnemo) [reply app/src/components/CreateLecturePopUp.tsx:117]: This code mesed with one of our PBI's, and was commented to make it work for now. We are going to work on this next sprint

File: app/src/pages/Courses.tsx
Comment (magnemo) [reply app/src/pages/Courses.tsx:48]: Nah have fixed it

File: app/src/pages/ExerciseDetail.tsx
Comment (magnemo) [reply app/src/pages/ExerciseDetail.tsx:66]: Nah, we fixed it

File: app/src/interfaces/CourseDetail.ts
Comment (Gogo-k) [reply app/src/interfaces/CourseDetail.ts:14]: This change made to match the MODLES.md. :)

File: app/src/pages/CourseEdit.tsx
Comment (Gogo-k) [reply app/src/pages/CourseEdit.tsx:65]: We are expecting it to be finished at the end of this sprint.

File: app/src/pages/CourseEdit.tsx
Comment (Gogo-k) [reply app/src/pages/CourseEdit.tsx]: Fixed.

File: app/src/pages/CourseEdit.tsx
Comment (Gogo-k) [reply app/src/pages/CourseEdit.tsx:100]: Possible code for implementing a cover image in the CourseEdit page. A comment has been added to reflect that. Thanks!

File: app/src/pages/CourseEdit.tsx
Comment (Gogo-k) [reply app/src/pages/CourseEdit.tsx:142]: Ye!

File: app/src/pages/CourseEdit.tsx
Comment (Gogo-k) [reply app/src/pages/CourseEdit.tsx:154]: Code for further development of changing coverimage of a course. A TODO comment has been added.

File: app/src/pages/CourseEdit.tsx
Comment (Gogo-k) [reply app/src/pages/CourseEdit.tsx:292]: Old code has been removed, fixed.

File: app/src/pages/ExerciseDetail.tsx
Comment (Gogo-k) [reply app/src/pages/ExerciseDetail.tsx:69]: Will be re-implemented.

File: app/src/pages/SectionEdit.tsx
Comment (Gogo-k) [reply app/src/pages/SectionEdit.tsx:67]: Commented out since backend route is bad, but will be added in the next pull request.

File: app/src/pages/SectionEdit.tsx
Comment (Gogo-k) [reply app/src/pages/SectionEdit.tsx:125]: Once again. Commented out since backend route is bad, but will be added in the next pull request.

File: app/src/services/course.services.ts
Comment (Gogo-k) [reply app/src/services/course.services.ts:78]: Done!

File: app/src/services/course.services.ts
Comment (Gogo-k) [reply app/src/services/course.services.ts:94]: Nope, and it has been deleted now.

File: app/src/services/lecture.services.ts
Comment (Gogo-k) [reply app/src/services/lecture.services.ts:17]: Yes, removed.

File: app/src/components/CreateLecturePopUp.tsx
Comment (SlovenlyTF) [reply app/src/components/CreateLecturePopUp.tsx:100]: Yeah, good catch. It was a test of design that we decided against.

File: app/src/components/CreateLecturePopUp.tsx
Comment (SlovenlyTF) [reply app/src/components/CreateLecturePopUp.tsx]: We had a though with it, but found a better way to do it.
So yes, it should just be deleted.

File: app/src/services/lecture.services.ts
Comment (SlovenlyTF) [reply app/src/services/lecture.services.ts:17]: We'll fix it to put.
There is an endpoint, it's just in our backend pull request.

File: app/src/services/exercise.services.ts
Comment (magnemo) [reply app/src/services/exercise.services.ts:21]: Thanks for the info, we have fixed it❤

**Review by TheMOHD3886 — APPROVED** [review]
sinds code fellas, 10/10 👍

File: app/src/services/section.services.ts
Comment (pernilje) [reply app/src/services/section.services.ts]: This has now been changed to be a put

File: app/src/services/section.services.ts
Comment (pernilje) [reply app/src/services/section.services.ts]: Thanks for the reminder, we have fixed it

File: app/src/services/exercise.services.ts
Comment (Jatewo) [inline app/src/services/exercise.services.ts:20]: eid probably shouldn't be any

**Review by Jatewo — CHANGES_REQUESTED** [review]
### Good work
I have a couple of general comments om top of the more specific ones found below:
- Some of the files are quite long. Perhaps we could consider splitting them up (I won't hold the PR back for this though)
- There are quite a few problems with the URLs not following our API guidelines. Perhaps you could take a look at these if you're not sure about it.
- There are a few hanging TODOs and commented out code. Please just take a quick look at these to ensure that they're not supposed to be fixed before merge 👍 

Another note: I currently have trouble creating a course and get a 404. If this is a mistake on my part, please let me know 👍 

Other than that (and perhaps the comments) it looks like you have done a great job 😄

File: app/src/components/Exercise/CreateExercisePopUp.tsx
Comment (pernilje) [reply app/src/components/Exercise/CreateExercisePopUp.tsx:119]: It's from last year, we have deleted it

File: app/src/components/Exercise/CreateExercisePopUp.tsx
Comment (pernilje) [reply app/src/components/Exercise/CreateExercisePopUp.tsx:58]: The documentation now follows Doxygen

File: app/src/services/exercise.services.ts
Comment (pernilje) [reply app/src/services/exercise.services.ts:20]: Now it's string | undefined instead

File: app/src/pages/CourseEdit.tsx
Comment (pernilje) [reply app/src/pages/CourseEdit.tsx:44]: It was from last year, but now it's deleted

File: app/src/components/Exercise/dropZone.tsx
Comment (magnemo) [reply app/src/components/Exercise/dropZone.tsx:1]: It was an duplicate file

File: app/src/pages/Courses.tsx
Comment (magnemo) [reply app/src/pages/Courses.tsx:42]: Comment has been updated

File: app/src/pages/SectionEdit.tsx
Comment (magnemo) [reply app/src/pages/SectionEdit.tsx:60]: We have reworded it

**Review by Jatewo — APPROVED** [review]
## 👍

**Review by JackSMunn — APPROVED** [review]

**Pull request merged**
